Control waypoint updater

ajaffer opened the pull request:
Can you see if this works fine on your machine?

zacwitte committed:
updated to handle acceleration and deceleration for stoplights

zacwitte committed:
merged the DBW and the waypoint updater nodes

Commit:
Edits to Twist controller

* server.py: I had to skip images, in order to keep it computationally light, see: discussion: https://carnd.slack.com/archives/C6NVDVAQ3/p1506794739000078 Note: we might need to take out the skipping while running on Carla
* twist_controller.py: *Added max_throttle to 100%, but for parking lot we might need a smaller value, plus I am unable to install the latest version of simulator, so I could only test the highway.

I noticed that the waypoint speed that the waypoint updater thinks is
sending is not the actual command being received by the twist-controller.

`[INFO] [1523852264.450267]: At waypoint 500, wp speed 0.0
[INFO] [1523852264.456432]: linear_vel: 11.1111111111 angular_vel: 0.00372815311879

....

[INFO] [1523852264.461660]: At waypoint 500, wp speed 0.0
[INFO] [1523852264.492005]: linear_vel: 11.1111111111 angular_vel: 0.00371841380014`

Commit:
Turning off logging, causes delays.

zacwitte committed:
correctly set the speed of the waypoint

zacwitte committed:
fixed index error in waypoint speeds

Commit:
A few changes to the twist controller:

* Taking the foot off throttle while pressing on the brakes
* incrementing the brake amount by 200%
* whenever velocity higher than requested: applying brakes + taking foot off throttle

With these changes I see that the car is able to do the fake stop with less jerky stop/go behavior.

The stop/go happens because the car creeps forward automatically when there is no brake, even though there is no throttle. I will keep thinking about this and see if I can find a better solution, but for now let's see if this works fine for us.

zacwitte commented on ros/src/twist_controller/dbw_node.py:
```diff
@@ -53,27 +53,60 @@ def __init__(self):
-            #                                                     <any other argument you need>)
-            # if <dbw is enabled>:
-            #   self.publish(throttle, brake, steer)
+            if not None in (self.current_vel, self.linear_vel, self.angular_vel):
```
I think what you mean is `if None not in (...):`

ajaffer replied:
Sorry, I didn't know if there is a difference, let me know if there is one?

```python
>>> b = [1,2,3]
>>> a = [4,5,6]
>>> if None not in (a,b):
...     print(a,b)
...
[4, 5, 6] [1, 2, 3]
>>> if not None in (a,b):
...     print(a,b)
...
[4, 5, 6] [1, 2, 3]
>>>


>>> a = None
>>> if not None in (a,b):
...     print(a,b)
...
>>> if None not in (a,b):
...     print(a,b)
...
>>>
```

zacwitte replied:
True, they're equivalent. I usually write `x not in y` because it's more clear. If you write not y in x it might be unclear whether you meant not (y in x) or (not y) in x. No need to change though. Just a stylistic nit.

zacwitte commented on ros/src/twist_controller/dbw_node.py:
```diff
@@ -53,27 +53,60 @@ def __init__(self):
+        self.current_vel = None
+        self.current_ang_vel = None
+        self.dbw_enabled = None
+        self.linear_vel = None
```
The name is this is ambiguous because it could be the current linear velocity or the desired linear velocity.

ajaffer replied:
fixed.

zacwitte commented on ros/src/twist_controller/dbw_node.py:
```diff
@@ -53,27 +53,60 @@ def __init__(self):
+        rospy.Subscriber('/current_velocity', TwistStamped, self.velocity_cb)
+        rospy.Subscriber('/vehicle/dbw_enabled', Bool, self.dbw_enabled_cb)
+
+        self.current_vel = None
```
A little documentation describing what each of these variables means would be helpful.

ajaffer replied:
done.

zacwitte commented on ros/src/twist_controller/twist_controller.py:
```diff
@@ -2,13 +2,108 @@
+    def __init__(self, vehicle_mass, fuel_capacity, brake_deadband, decel_limit,
+        accel_limit, wheel_radius, wheel_base, steer_ratio, max_lat_accel, max_steer_angle):
+
+        self.yaw_controller = YawController(wheel_base, steer_ratio, 0.1, max_lat_accel, max_steer_angle)
```
What is the `0.1` value?

ajaffer replied:
added doc.

zacwitte commented on ros/src/twist_controller/dbw_node.py:
```diff
@@ -53,27 +53,60 @@ def __init__(self):
+              self.publish(self.throttle, self.brake, self.steering)
             rate.sleep()
 
+    def twist_cb(self, msg):
```
Some documentation on each of these callbacks describing what they mean would be helpful.

ajaffer replied:
done.

zacwitte commented on ros/src/twist_controller/dbw_node.py:
```diff
@@ -53,27 +53,60 @@ def __init__(self):
-            # if <dbw is enabled>:
-            #   self.publish(throttle, brake, steer)
+            if not None in (self.current_vel, self.linear_vel, self.angular_vel):
+                    self.throttle, self.brake, self.steering = self.controller.control(self.current_vel,
```
The way the code is formatted here is a little difficult to read. Indenting the params of the `control()` func would help.

ajaffer replied:
fixed.

zacwitte commented on ros/src/twist_controller/twist_controller.py:
```diff
@@ -2,13 +2,108 @@
+        # self.max_throttle = 0.2
+        self.max_throttle = 1.0
+
+        kp = 0.3
```
How did you arrive at these values? Can we tweak them to reduce the start-stop behavior at low speeds?

ajaffer replied:
I started with the values from lecture notes, experimented with different set of values using the ROS bag reference implementation and plotted the graphs of brakes, throttles and steering, these were the best values I settled on. The car also drives good using these values, we can tweak them if needed. These are no listed in styx.launch file as params.

zacwitte commented on ros/src/twist_controller/twist_controller.py:
```diff
@@ -2,13 +2,108 @@
+            self.throttle_controller.reset()
+            return 0., 0., 0.
+
+        current_vel = self.vel_lpf.filt(current_vel)
```
A comment about what this line doe would be helpful.

ajaffer replied:
Added doc.

zacwitte commented on ros/src/twist_controller/twist_controller.py:
```diff
@@ -2,13 +2,108 @@
+
+        throttle = self.throttle_controller.step(vel_error, sample_time)
+        brake = 0
+        if linear_vel == 0. and current_vel < 2.0:
```
Not clear what `linear_vel` and `current_vel` mean here. Could this be the source of the stop-start at low speeds?

ajaffer replied:
fixed the confusion with variable naming, but this is not the cause of the low speed stop-start behavior

zacwitte commented on ros/src/twist_controller/twist_controller.py:
```diff
@@ -2,13 +2,108 @@
+        brake = 0
+        if linear_vel == 0. and current_vel < 2.0:
+            throttle = 0
+            brake = 400
```
is `400` an arbitrary number?

ajaffer replied:
This is for a full stop.

ajaffer replied:
See my latest review, when the car is at low speeds and the desired speed is close to a complete stop: let go of the throttle and apply a constant Torque of 400 N . m, to keep the car from creeping forward.

zacwitte reviewed (changes requested):
Mostly questions and comments about code clarity. We should hunt down the source of the start-start behavior at low speeds before we merge this in.

zacwitte committed:
trigger green light after stopping at red light

zacwitte committed:
merged master

zacwitte committed:
merged master

Commit:
Achieved much smoother stop behavior + changes for comments

Commit:
fix typo

Commit:
Successfully able to stop and re-start the car.

Car stops much smoother now, let me know if there are issues still.

twist_controller.py
-------------------
* Using a separate PID controller for brake
* Added a low pass filter for steering
* Cleaned up the brake/throttle logic

waypoint_updater.py
-------------------
* Using a linear extrapolation for the desired acceleration and deceleration velocities in the waypoint_updater. The issue was that if both waypoint_updater and twist_controller are plotted on a curve, that means a curve is trying to converge onto a curve. It is much better if a curve is trying to converge on a straight line, and it is proven that this works better.
* Increased the deceleration distance to 30 m, gives a better spread out waypoints
* changed the target deceleration velocity to a negative, that gives the last two points as zero (or below zero, which is essentially zero), which gives more chances for the car to hit the target waypoint in order to stop

[Minor]
* dbw_node.py: Documentation
* styx.launch: Needed to skip images, otherwise my CPU can't keep up while rendering the dense tree section

ajaffer commented on ros/src/twist_controller/twist_controller.py:
```diff
@@ -58,49 +69,24 @@ def control(self, current_vel, current_ang_vel, dbw_enabled, desired_linear_vel,
+            # rospy.logwarn('complete stop')
+        elif (vel_error < 0.0):
+            decel = max(vel_error, self.decel_limit)
+            decel = self.brake_controller.step(abs(decel), sample_time)
```
**Previous Stop-Go Issue**
The following picture shows the prev stop-go behavior, here you can see the car's current velocity (blue) catches up with the desired velocity (red) three times, hence causing the stop-go behavior:

<img width="1200" alt="screen shot 2018-05-03 at 3 11 46 am" src="https://user-images.githubusercontent.com/5813/39573822-a49689b0-4e89-11e8-86f5-87b1877c301a.png">



**With this Fix**
With the PID Controller for brake and other fixes in this commit, the graph is much smoother, here it shows how the current velocity (blue) catches up to the desired velocity (red).

<img width="1161" alt="screen shot 2018-05-03 at 4 19 36 am" src="https://user-images.githubusercontent.com/5813/39573725-3e28e42a-4e89-11e8-820e-48b0c75da678.png">

ajaffer commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -163,7 +164,7 @@ def traffic_cb(self, msg):
         # print(x)
         # print(y)
-        F = interp1d(x, y, kind='cubic')
+        F = interp1d(x, y)
```
<img width="645" alt="screen shot 2018-05-03 at 3 52 13 am" src="https://user-images.githubusercontent.com/5813/39573765-6497c054-4e89-11e8-9bfe-b9cec8230cdd.png">
<img width="642" alt="screen shot 2018-05-03 at 3 50 28 am" src="https://user-images.githubusercontent.com/5813/39573766-64b1670c-4e89-11e8-8b44-415c6d97af18.png">

Interpolating linearly for both desired acceleration and deceleration distance/velocities. See commit description for details.

anandman committed:
commented out classifier code until we get it working

anandman committed:
removed fake starts and stops

anandman committed:
put tl_detector back in
